src/lightning/data/datasets/iterable.py

```diff
@@ -0,0 +1,447 @@
+        """Prepares a single chunk before it is actually loaded. This could e.g. download the actual file from s3.
+
+        Args:
+            chunk: the chunk data to prepare.
+        """
+
+    @abstractmethod
+    def load_chunk(self, chunk: Any) -> Any:
+        """Loads a single chunk into memory. This could e.g. mean loading the file that has previously been
+        downloaded from s3.
+
+        Args:
+            chunk: The chunk that should be currently loaded
+        """
+
+    @abstractmethod
+    def load_sample_from_chunk(self, chunk: Any, index: int) -> Any:
+        """Retrieves a single sample from a given (already loaded) chunk. This could be indexing a list or
+        returning the entire chunk if it's size is 1.
+
+        Args:
+            chunk: The chunk the sample should be retrieved from
+            index: The index of the current sample to retrieve within the chunk.
+        """
+
+    def __iter__(self) -> "LightningIterableDataset":
+        """Creates an iterator.
+
+        Before that, detects the env if necessary, shuffles chunks, shards the data and shuffles sample orders within
+        chunks.
+        """
+        self._curr_chunk_index = self._start_index_chunk
+        self._curr_sample_index = self._start_index_sample
+        if self._env.worker_env is None:
+            self._env.worker_env = WorkerEnv.detect()
+
+        self._chunks = self._shuffle_if_necessary(self._chunks, 0, shuffle_chunk_order=True, shuffle_sample_order=False)
+        self._apply_sharding()
+        self._local_chunks = self._shuffle_if_necessary(
+            self._local_chunks,
+            self._curr_chunk_index,
+            shuffle_chunk_order=False,
+            shuffle_sample_order=True,
+        )
+        self._ensure_chunks_loaded()
+        return self
+
+    def __next__(self) -> Any:
+        """Returns the next sample.
+
+        If necessary, this also loads the new chunks.
+        """
+        try:
+            self._check_if_sharded()
+            self._ensure_chunks_loaded()
+
+            if self._curr_sample_index >= self._curr_loaded_num_samples:
+                self._curr_chunk_index += self._num_parallel_chunks
+                self._check_dataset_end()
+
+                self._load_next_chunks()
+                self._curr_sample_index = 0
+
+            remainder = self._curr_sample_index
+            curr_loaded_chunk_idx = 0
+            for i, c in enumerate(self._curr_loaded_chunks):
+                if c.chunk_size > remainder:
+                    curr_loaded_chunk_idx = i
+                    break
+
+                remainder -= c.chunk_size
+
+            sample = self.load_sample_from_chunk(
+                self._curr_loaded_chunks[curr_loaded_chunk_idx]._chunk_data,
+                self._curr_loaded_chunks[curr_loaded_chunk_idx].index_permutations[remainder],
+            )
+
+            return sample
+
+        finally:
+            self._curr_sample_index += 1
```

Comment:
Could you remind me why is it correct to increment the index when failing on the current one? Wouldn't we want to "resume" on the last incomplete sample?

Reply:
We always skip to the next full chunk. I can change the `finally` to do it immediately after the sample is loaded.